### ISSUE TITLE
Need some context information in the `OSSL_FUNC_keymgmt_import/export_types`

### ISSUE THREAD

**Issue by afshinpir** [openssl/openssl#20479]
Hi

Since I'm writing my own provider, I have encountered an strange limitation in `OSSL_FUNC_keymgmt_import_types` and `OSSL_FUNC_keymgmt_export_types`. These methods only receive `selection` as argument. The problem is that this is not enough information to find our context if we don't want to write a lot of duplicate code. From outside point of view, it is obvious what these types are for, but it is not easy to implement them without a lot of code bloat. 
As an example, let's assume that in our keymgmt, we want to handle importing both `RSA` and `ECC` keys. The only feasible method right now if to write a different set of methods for `RSA` and `ECC`:
```
const OSSL_DISPATCH l_RsaKeymgmtDispathTable[] = { \
    //....
    { OSSL_FUNC_KEYMGMT_IMPORT_TYPES, (void (*)(void))RsaKeyImportTypes }, \
    { OSSL_FUNC_KEYMGMT_EXPORT_TYPES, (void (*)(void))RsaKeyExportTypes }, \
    { 0, nullptr } \
};
const OSSL_DISPATCH l_EccKeymgmtDispathTable[] = { \
    //....
    { OSSL_FUNC_KEYMGMT_IMPORT_TYPES, (void (*)(void))EccKeyImportTypes }, \
    { OSSL_FUNC_KEYMGMT_EXPORT_TYPES, (void (*)(void))EccKeyExportTypes }, \
    { 0, nullptr } \
};
const OSSL_ALGORITHM l_KeymgmtDispathTable[] = {
    ALG(GGL_PROV_NAME_RSA, l_RsaKeymgmtDispathTable),
    ALG(GGL_PROV_NAME_ECC, l_EccKeymgmtDispathTable),
    ALGS_END
};
```
It is obvious `RsaKeyImportTypes` is `RSA` and `EccKeyImportTypes` an `Ecc` context if we write the code like this, but we cannot write a single method that handle both key types like this now:
```
const OSSL_DISPATCH l_RsaEccKeymgmtDispathTable[] = { \
    //....
    { OSSL_FUNC_KEYMGMT_IMPORT_TYPES, (void (*)(void))RsaEccKeyImportTypes }, \
    { OSSL_FUNC_KEYMGMT_EXPORT_TYPES, (void (*)(void))RsaEccKeyExportTypes }, \
    { 0, nullptr } \
};
const OSSL_ALGORITHM l_KeymgmtDispathTable[] = {
    ALG(GGL_PROV_NAME_RSA, l_RsaEccKeymgmtDispathTable),
    ALG(GGL_PROV_NAME_ECC, l_RsaEccKeymgmtDispathTable),
    ALGS_END
};
```
Because we don't get anything about context in in(ex)port_types methods. I hope we can have something which can be used to make distinction between contexts. I should say that I have checked the code around here and it is feasible to add `key_type` in these methods syntax, so becomes something like:
```
const OSSL_PARAM *OSSL_FUNC_keymgmt_import_types(int selection, const char* key_type);
```
where `key_type` is the name that we set when calling `EVP_PKEY_new_from_name`.
In addition, I should say that there has been an update to these functions last week with `_ex` functions which provide provider context, but even with provider context we cannot make this distinction.
I probably can make a PR for this feature, but I wanted to know everyone's opinion first.

**Comment by slontis** [comment]
@levitte ?? (Not sure it was to share different key type in the same keymanager)

**Comment by levitte** [comment]
By design, each provider implementation is expected to handle a single thing, and certain things are expected to be fairly static.  This means that each keymgmt implementation is expected to be about one key type.  It has been done this way so that, when an implementation has been fetched, there's no question what it is about.

Also, you might want to be aware that the key type that would be passed to your example version of `OSSL_FUNC_keymgmt_import_types()` isn't necessarily one unique value, as it's possible an alias could be passed.

In practice, this only really holds true with the OSSL_ALGORITHM and OSSL_DISPATCH tables and the functions they point to.  Those function may then call other functions with extra parameters that indicate a desired key type, for example, so you may merge similar functionalities as you wish, "under the hood".

**Comment by levitte** [comment]
As this is a feature request, I might respond with regards to that.

Quite frankly, I'm not sure at all about the wisdom regarding passing the key type as a parameter these functions, because I believe in the clarity of what was fetched.  I don't think that the ramification of having a OSSL_DISPATCH denote multiple key types, which would lead to the question "what did I actually fetch, really?", would be very fruitful.

That being said, we have had the wish that we could pass a keymgmt implementation context to functions like `OSSL_FUNC_keymgmt_import_types`, 'cause there have been times when the answer could possibly maybe vary depending on the current state of a fetched keymgmt instance.

**Comment by t8m** [comment]
Please look at ecx_keymgmt.c how the code duplication is mostly avoided. Yes, you'll need a thin wrapper function for all the functions where you need to distinguish the key type but IMO that is fine.

**Comment by ifranzki** [comment]
Take a look at https://github.com/openssl/openssl/commit/5e3b84505e44377b183e7529dab7585674b83936, this might help you as well.

**Comment by nhorman** [comment]
marking as inactive, will be closed at the completion of 3.4 dev without further input

**Comment by t8m** [comment]
I am closing this as there is nothing here to fix/implement.
